### ISSUE TITLE
How can I use AIMET with custom ops?

### ISSUE THREAD

**Issue by StephenChou0119** [quic/aimet#2574]
I'm using AutoQuantV2 in AIMET 1.28 with pytorch 1.13.1. My model have a custom ops named bev pooling v2, which is introduced by BEVDET. 

1. If I wrap my custom torch extension with  torch.autograd.Function. The presence of torch.autograd.Function in the model will be interpreted as PythonOP, preventing the generation of a quantized model.
2. If I use register_custom_op_symbolic and wrap torch.ops.custom_ops.bev_pool_v2, the model prepare cannot pass even I wrap it with torch.nn.Module.
3.  However, a CPP extension wrapped with torch.fx.wrap (not torch.ops function) can be ignored at model prepare phrase.  It can pass model prepare and batch norm, but cpp extension can not be exported to onnx.

### code 1
```
class BevPoolWrappedFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, depth, img_feat, 
                                    ranks_bev,
                                    ranks_depth, ranks_feat, interval_starts, 
                                    interval_lengths, 
                                    # drop_ranks,
                                    # n_intervals,
                                    batch_size, 
                                    z_size, y_size, x_size, channel_num):
        out = torch.zeros(batch_size, y_size, x_size, channel_num, dtype=torch.float32, device="cuda:0")
        bev_pool_ext.bev_pool_v2_forward(
            depth,
            img_feat,
            out,
            ranks_depth,
            ranks_feat,
            ranks_bev,
            interval_lengths,
            interval_starts,
        )
        return out

    @staticmethod
    def symbolic(g, depth, img_feat, 
                                    ranks_bev,
                                    ranks_depth, ranks_feat, interval_starts, 
                                    interval_lengths, 
                                    # drop_ranks,
                                    # n_intervals,
                                    batch_size, 
                                    z_size, y_size, x_size, channel_num):
        return g.op("custom_domain::bev_pool_v2_forward_quantization", depth, img_feat, ranks_bev, ranks_depth, 
                            ranks_feat, interval_starts, interval_lengths,
                            # drop_ranks,
                            # n_intervals,
                             batch_size_i = batch_size, z_size_i = z_size, y_size_i = y_size, x_size_i = x_size, channel_num_i = channel_num,)
```
### code 2
```
@parse_args('v', 'v', 'v', 'v', 'v', 'v', 'v', 'i', 'i', 'i', 'i', 'i')
def symbolic_bev_pool_v2_forward_quantization(g, depth, img_feat, 
                                    ranks_bev,
                                    ranks_depth, ranks_feat, interval_starts, 
                                    interval_lengths, 
                                    # drop_ranks,
                                    # n_intervals,
                                    batch_size, 
                                    z_size, y_size, x_size, channel_num):
    return g.op("custom_domain::bev_pool_v2_forward_quantization", depth, img_feat, ranks_bev, ranks_depth, 
                            ranks_feat, interval_starts, interval_lengths,
                            # drop_ranks,
                            # n_intervals,
                             batch_size_i = batch_size, z_size_i = z_size, y_size_i = y_size, x_size_i = x_size, channel_num_i = channel_num,  outputs=1)
register_custom_op_symbolic("custom_ops::bev_pool_v2_forward_quantization", symbolic_bev_pool_v2_forward_quantization, 14)
class BevPoolModuleOps(torch.nn.Module):
    def forward(self, out, depth, feat, ranks_bev, ranks_depth, ranks_feat, interval_starts, interval_lengths, batch_size, z_size, y_size, x_size, channel_num):
        depth = depth.contiguous().float()
        img_feat = feat.contiguous().float()
        ranks_bev = ranks_bev.int()
        ranks_depth = ranks_depth.contiguous().int()
        ranks_feat = ranks_feat.contiguous().int()
        interval_starts = interval_starts.contiguous().int()
        interval_lengths = interval_lengths.contiguous().int()
        out = torch.ops.custom_ops.bev_pool_v2_forward_quantization(depth, img_feat, 
                                        ranks_bev,
                                        ranks_depth, ranks_feat, interval_starts, 
                                        interval_lengths,
                                        batch_size, 
                                        z_size, y_size, x_size, channel_num)
        out = out.permute(0, 3, 1, 2,).contiguous()
        return out
```

```
Consider the following choices: 
1. Redefine as a torch.nn.Module in the class definition.
2. The op can remain as a functional op due to being math invariant, but the op type has not been added to ConnectedGraph.math_invariant_types set. 
Add an entry to ignore the op by importing the set and adding the op type:

        from aimet_torch.meta.connectedgraph import ConnectedGraph
        ConnectedGraph.math_invariant_types.add(...)

The following functional ops were found. The parent module is named for ease of locating the ops within the model definition.
        bev_pool_v2_forward_quantization_359          parent module: GraphModule
```

### code 3
```
@torch.fx.wrap
def bev_pool_wrapped(out, depth, img_feat, 
                                    ranks_bev,
                                    ranks_depth, ranks_feat, interval_starts, 
                                    interval_lengths, 
                                    # drop_ranks,
                                    # n_intervals,
                                    batch_size, 
                                    z_size, y_size, x_size, channel_num):
    depth = depth.contiguous().float()
    img_feat = img_feat.contiguous().float()
    ranks_bev = ranks_bev.int()
    ranks_depth = ranks_depth.contiguous().int()
    ranks_feat = ranks_feat.contiguous().int()
    interval_starts = interval_starts.contiguous().int()
    interval_lengths = interval_lengths.contiguous().int()
    bev_pool_ext.bev_pool_v2_forward(
        depth,
        img_feat,
        out,
        ranks_depth,
        ranks_feat,
        ranks_bev,
        interval_lengths,
        interval_starts,
    )
    return out
class BevPoolModule(torch.nn.Module):
    def forward(self, out, depth, feat, ranks_bev, ranks_depth, ranks_feat, interval_starts, interval_lengths, batch_size, z_size, y_size, x_size, channel_num):
        out = bev_pool_wrapped(out, depth, feat, ranks_bev, ranks_depth, ranks_feat, interval_starts, interval_lengths, batch_size, z_size, y_size, x_size, channel_num)
        out = out.permute(0, 4, 1, 2, 3).contiguous()
        out = out.squeeze(2)
        return out
```

**Comment by quic-ashvkuma** [comment]
The second workflow that you outlined should be supported in AIMET 1.30. Specifically, it will be possible to create a PyTorch model using a custom op registered with `torch.onnx.register_custom_op_symbolic`. To take advantage of this feature, you will have to export your model (with custom op) to ONNX, and create a quantized model using this exported model and AIMET ONNX.

**Comment by StephenChou0119** [comment]
I am using MQBench Now, questions above were worked around.

**Comment by quic-muchhsu** [comment]
Looks like it is solved. Closing.
